File: v1/src/main/java/com/api/v1/persons/services/PersonRegistrationServiceImpl.java
package com.api.v1.persons.services;

import com.api.v1.persons.domain.Person;
import com.api.v1.persons.domain.PersonRepository;
import com.api.v1.persons.dtos.PersonRegistrationRequestDto;
import com.api.v1.persons.exceptions.DuplicatedEmailException;
import com.api.v1.persons.exceptions.DuplicatedSsnException;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class PersonRegistrationServiceImpl implements PersonRegistrationService {

    @Autowired
    private PersonRepository personRepository;

    @Override
    public Mono<Person> register(@Valid PersonRegistrationRequestDto requestDto) {
        return personRepository
                .findBySsn(requestDto.ssn())
                .hasElement()
                .flatMap(ssnExists -> {
                    if (ssnExists) return Mono.error(DuplicatedSsnException::new);
                    return personRepository
                            .findByEmail(requestDto.email())
                            .hasElement()
                            .flatMap(emailExists -> {
                                if (emailExists) return Mono.error(DuplicatedEmailException::new);
                                return personRepository.save(Person.of(requestDto));
                            });
                });
    }

}

File: kotlin/v3/src/main/kotlin/com/api/v3/persons/services/PersonRegistrationServiceImpl.kt
package com.api.v3.persons.services

import com.api.v3.persons.domain.Person
import com.api.v3.persons.dtos.PersonRegistrationRequestDto
import com.api.v3.persons.domain.PersonRepository
import com.api.v3.persons.exceptions.DuplicatedEmailException
import com.api.v3.persons.exceptions.DuplicatedSsnException
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Service

@Service
internal class PersonRegistrationServiceImpl(
    private val personRepository: PersonRepository
): PersonRegistrationService {

    override suspend fun register(requestDto: @Valid PersonRegistrationRequestDto): Person {
        return withContext(Dispatchers.IO) {
            isGivenSsnDuplicated(requestDto.ssn)
            isGivenEmailDuplicated(requestDto.email)
            personRepository.save(Person.of(requestDto))
        }
    }

    private suspend fun isGivenSsnDuplicated(ssn: String) {
        if (personRepository.findBySsn(ssn) != null) {
            throw DuplicatedSsnException()
        }
    }

    private suspend fun isGivenEmailDuplicated(email: String) {
        if (personRepository.findByEmail(email) != null) {
            throw DuplicatedEmailException()
        }
    }

}